ext/socket/lib/socket.rb

```diff
@@ -593,6 +594,227 @@ def accept_nonblock(exception: true)
     __accept_nonblock(exception)
   end
 
+  def self.start_getaddrinfo_v6(host, port, mutex, pipe_write, ai_queue_v6, error_queue)
+    return Thread.new do
+      Thread.current.report_on_exception = false
+      ai_list = Addrinfo.getaddrinfo(host, port, :PF_INET6, :STREAM)
+      ai_list.each {|ai| ai_queue_v6.push(ai) }
+    rescue SocketError => ex
+      case ex.message
+      when "getaddrinfo: Address family for hostname not supported" # when IPv6 is not supported
+        # ignore
+      when "getaddrinfo: Temporary failure in name resolution" # when timed out (EAI_AGAIN)
+        # ignore
+      else
+        error_queue.push(ex) # report the exception to main thread
+      end
+    rescue => ex
+      error_queue.push(ex)
+    ensure
+      mutex.synchronize { pipe_write.puts("getaddrinfo_v6_done") unless pipe_write.closed? }
+    end
+  end
+  private_class_method :start_getaddrinfo_v6
+
+  # 50ms is the recommended value for the resolution delay for IPv4 in RFC8305
+  RESOLUTION_DELAY = 0.05
+  private_constant :RESOLUTION_DELAY
+
+  def self.start_getaddrinfo_v4(host, port, mutex, pipe_write, ai_queue_v4, error_queue)
+    return Thread.new do
+      Thread.current.report_on_exception = false
+      begin
+        ai_list = Addrinfo.getaddrinfo(host, port, :PF_INET, :STREAM)
+      rescue SocketError => ex
+        case ex.message
+        when "getaddrinfo: Address family for hostname not supported" # when IPv4 is not supported
+          # ignore
+        when "getaddrinfo: Temporary failure in name resolution" # when timed out (EAI_AGAIN)
+          # ignore
+        else
+          error_queue.push(ex) # report the exception to main thread
+        end
+      rescue => ex
+        error_queue.push(ex)
+      ensure
+        mutex.synchronize { pipe_write.puts("getaddrinfo_v4_done") unless pipe_write.closed? }
+      end
+      if ai_list # if getaddrinfo finished successfully
+        ai_list.each {|ai| ai_queue_v4.push(ai) }
```

Comment:
It sends addrinfos after "getaddrinfo_v4_done".
So, there is a (small) moment that "getaddrinfo_v4_done" is sent and addrinfos are not sent.

The loop condition in make_connection_attempts is
`while !getaddrinfo_v6_done || !getaddrinfo_v4_done || !ai_queue_v6.empty? || !ai_queue_v4.empty?`.
I guess there is rase condition here: `!getaddrinfo_v4_done || !ai_queue_v4.empty?` will be false at the moment.
Assuming the server has no IPv6 address, it finishes the loop prematurely and IPv4 addresses will be ignored.

I think the protocol between the getaddrinfo-thread and main-thread should be improved.
I expect that the protocol should be changed as:
when the main-thread detected a notification from the getaddrinfo-thread via the pipe, the main-thread obtain the result as a single object from the queue.